### ISSUE TITLE
distsql: OOM with 3-way TPC-H query 3

### ISSUE THREAD

**Issue by cuongdo** [cockroachdb/cockroach#15194]
Build: 1e0b89a
Config: 6-node cluster Standard_D3 Azure VMs
Test: `tpch -concurrency 3 -queries 3`

memprof:
[profile001.pdf](https://github.com/cockroachdb/cockroach/files/939440/profile001.pdf)

Logs:
[cockroach.stderr.txt](https://github.com/cockroachdb/cockroach/files/939446/cockroach.stderr.txt)

[DistSQL query plan](https://cockroachdb.github.io/distsqlplan/decode.html?eJzcmFFv4kYUhd_7K6x52aSdSMyMgayllSbaTbtUDVuR9KGq0IrFU2KJYGQbqatV_nuFWS1g03t2fCFA35zA5-N753KOZ76IWRq7_ujJ5SL6SyghhRZShEKKtpCiI4ZSzLN07PI8zZZfWQG9-B8RtaRIZvNFsfz3UIpxmjkRfRFFUkydiMTD6NPUDdwodpmQInbFKJmWIvMseRpln-14kRfpU_npz8m0cFkU2G7wJnh1d_P2fa9_O_jzVRRF9w-DXv8XIcWHRREFVknbFcNnKdJF8VV7Lfnpc_A4yh-35awSw-fhs9zDk08_5rGdJjOXFO5p_UjVB_IS-zVNZlSX6nLSdvbdhDW9mKVZ7DIXb_Elib6yo7j3o_xxWWCtke7v4sKqyzdZMnksr76Vp6U10obStv-vdQ7cLC6HvbOc9x-DCxVF0bvbt727m9-Cq8BeXwY2DGy7_MLygjNeN5NJ5iajIq082bf2StF7d9t_-Pps93_cXVh9uf6v2bgON67bG9edjevupeeCrZeZs2yHr5yxBvdpVlQnw-orac1P333b9d12NWDjdjv0--lVOq_O-X-qqpc0RyB2NHNUp2YaBzLHk6tzv-YIxutszdFv2Y5rjmANmpqjbmhT6cc0tmWf8rVrtDhSyKSqYnL5M25Le83-9TbtQSOrZnbhYFatT83CDmTVJ1fnfq0ajNfZWrXfsh3XqsEaNLVq85I2BcSOZlPm1H6-B7Kpk6tzvzYFxutsbcpv2Y5rU2ANmtpUeLbnmU2fvMG7MJA63rtw0x40ChlmFw4WMuGpme-BQubk6txvyIDxOtuQ8Vu244YMWIN9nOnuuO3A5fN0ljvytHYohYsnbtXFPF1kY_d7lo7Lm6_-_FB-uzx7iV1erD4NV3_0ZquPlo_1_bC55tBasWiWttI0rap0a0t7C25VYU03rbtFaz9at1g0S1spDm1o2HiMqSdcGVNPWoOqAc3SroxpjQ7JMW3TY9ompTv0D4SGwxaH1oZDG5a2AobYITvepTveJWFgaNckHAJPek3SqkXTCuQAsBWAI09DOE8duRqqHdC1JPHxNUAjYwM4cjaE89SRt6lanGyPrAEjSzsrsDdAI38DODI4gCOHAziyOEWnigKxomoG6-NyijZYBRxW1SyWMqq6Oo1r1HiA89QVeF9GtQOazhf1GjS-FjB-CUHjMCEAzlOHCQFqB2_NPhs0XxolBMBRQiCcp44SQoNdGtqm0fsVkBCARgkBcJQQAEcJAXCUELoWMNuND0Hj6YABCaHpfNEdIF4LGL_dOY3D7TnAeepwgw5qBzSdL8ioQDohqwDRiioHOE8dGhWdrRpkq6bzBRkVTUOjAtGKjApkI08dGZWpBcxW442iG2_ogAFGZUBCsGh09ApwdPaKcJ46On01dL4YkC-mli9emwCAo00AwnnqaBOAagc0na0GZKvhZSvAUbYinKeOshXVDmhWtgIann_zshXhPHWUrYbOVgOy1bCyFdAoWwGOshXgKFsBjrI1pLM1BNka-mXr8PmHfwMAAP__ydZW-g==)

**Comment by cuongdo** [comment]
query is here: https://github.com/cockroachdb/loadgen/blob/master/tpch/query3.go

**Comment by knz** [comment]
I think this is addressed by Justin's memory PR.

**Comment by cuongdo** [comment]
@justinj says his PR unfortunately doesn't fix this

**Comment by justinj** [comment]
The only change my PR should affect in DistSQL is use of `CONCAT_AGG` and `ARRAY_AGG` so I don't think it will address this, but it does make memory monitoring more accessible to DistSQL so it would conceivably allow us to track what's happening here.

**Comment by RaduBerinde** [comment]
In addition to memory accounting (which would turn the OOM into query errors), we can save memory by never storing `EncDatum`s (e.g. in `sorter`), and storing just `Datum`s instead. The `EncDatum`s are larger, and (more importantly), their encodings can keep bigger slices alive.

We should also fix the group renders to avoid creating duplicates. In this plan we have three duplicate renders before the aggregation (and we store their data for each group).

**Comment by petermattis** [comment]
Did the recent work on memory accounting fix the OOM here?

**Comment by arjunravinarayan** [comment]
We should attempt to reproduce this, which I can do once rc2.2 builds. From the number of rows in the lineitem table, it looks like scalefactor=10 was used.

**Comment by arjunravinarayan** [comment]
I am now using `denim` to attempt to reproduce this. I will check back in in a few hours once we've given it an opportunity to OOM.

**Comment by arjunravinarayan** [comment]
Running this load generator configuration crashed the cockroach node (denim-4, logs are saved), and the load generator exited with the following output:
`2017/05/04 17:35:20 [1] error running query 3: pq: [n6] communication error: rpc error: code = Canceled desc = context canceled`

It looks like an Out of Memory error to me. The node has not been restarted, the logs directory on denim-4 is as of the crash, but I've also tarballed it as `~/tpch.query3.oom.logs.tgz` just in case.

**Comment by arjunravinarayan** [comment]
@RaduBerinde do you want to take a look? I have to step out for a short bit.

**Comment by RaduBerinde** [comment]
A memory profile would help if indeed it is OOM.

**Comment by arjunravinarayan** [comment]
That is saved in the logs

**Comment by arjunravinarayan** [comment]
Update: turns out `denim` was configured to not save memory profiles for performance reasons (it is the cluster where we run benchmarks), but I was misled by the fact that the memory profile _files_ were in the logs directory - however, they were empty upon closer inspection due to memory profiling being disabled.

Thus, we needed to rerun this experiment, after restarting cockroach with an updated supervisor config configured to save non-empty memory profiles. Therefore I upgraded `denim` to the current rc2 candidate (SHA `1700211837c3e519bf093f8cda26fe2afb028ff3`) with the updated configuration, and reran `tpch -concurrency 3 -queries 3` on denim 7.

Except, this time, the cluster is _not_ OOMing, and the queries are completing (albeit at a runtime of 40 minutes per query!). So we're in wait and watch mode until it repros.

**Comment by RaduBerinde** [comment]
What build was used previously? (wondering if it included my latest mem accounting changes)

**Comment by arjunravinarayan** [comment]
The build used was `32862bd08f4fdb202475eb6eccc30bff38598e11`.

Curiously, the connection to the load generator was closed:
```
2017/05/04 21:45:48 [1] finished query 3: 113376 rows returned after 2454.32 seconds
2017/05/04 21:57:52 [0] finished query 3: 113376 rows returned after 3178.51 seconds
2017/05/04 21:58:19 [2] finished query 3: 113376 rows returned after 3204.72 seconds
2017/05/05 00:45:49 [1] error running query 3: driver: bad connection
```

It's probably HAProxy killing the connection, as the connection timeouts are set to 1 hour. From the growth in the execution time, it might have passed the 3600 mark on the next query. I've reset HAProxy's timeouts to 6h, and restarted the jobs. I suspect that we may no longer be OOMing on this SHA, we will know tomorrow morning.

**Comment by arjunravinarayan** [comment]
Post-weekend update: late on Thursday we OOMed on 3 machines (`denim 1,3,5`), without completing a single query. The build used was `1700211837c3e519bf093f8cda26fe2afb028ff3 `, the same as last time. Investigating now.

**Comment by arjunravinarayan** [comment]
We confirmed from the supervisor log and `dmesg` that it was in fact an out of memory error. We proceeded to examine the memory profile which was saved since I turned it on  (thanks to @tschottdorf for helping @asubiotto and me debug this 🎉 ).

TL;DR: We need memory accounting in `HashJoiner`, since that's what's blowing up. @asubiotto is doing this today.
---

The following log of what I found as I'm new to memory profiling in go, and want to maintain some records that I can reference later.

The `top10` memory users pointed to a pretty solid lead:
```
(pprof) top10
8193.67MB of 8519.95MB total (96.17%)
Dropped 1465 nodes (cum <= 42.60MB)
Showing top 10 nodes out of 44 (cum >= 360.54MB)
      flat  flat%   sum%        cum   cum%
 5420.02MB 63.62% 63.62%  5422.02MB 63.64%  github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DecodeTableValue
 1343.05MB 15.76% 79.38%  1343.05MB 15.76%  github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowContainer).allocChunks
  924.51MB 10.85% 90.23%   924.51MB 10.85%  runtime.hashGrow
  326.52MB  3.83% 94.06%  7993.61MB 93.82%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*hashJoiner).buildPhase
      92MB  1.08% 95.14%       92MB  1.08%  runtime.rawstringtmp
   76.54MB   0.9% 96.04%  1010.06MB 11.86%  runtime.mapassign
    7.03MB 0.082% 96.12%   204.05MB  2.39%  github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*procOutputHelper).emitRow
       2MB 0.023% 96.15%    51.71MB  0.61%  github.com/cockroachdb/cockroach/pkg/storage.evaluateCommand
       1MB 0.012% 96.16%   139.26MB  1.63%  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processUnaryRPC
       1MB 0.012% 96.17%   360.54MB  4.23%  github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*EncDatum).Encode
```
Looking at the PDF, it's pretty clear what the culprit is: The `HashJoiner` is building a giant bucket of decoded tuples and never making progress, and eventually running out of memory: 
![hashjoineroom](https://cloud.githubusercontent.com/assets/1565833/25811536/9a123ad4-33e1-11e7-90d4-c3b721a4a595.png)

Therefore, we need memory accounting in the HashJoiner. @asubiotto is working on a PR to add this now.

**Comment by arjunravinarayan** [comment]
@asubiotto it sounds like you've made a bunch of progress on memory accounting, and that it's far more subtle than I thought. Do you mind giving an update? It seems like I was mistaken in thinking this was just the lack of memory accounting. The `HashJoiner` already has memory accounting, but one of the code paths might be mis-accounted - at least that was the last theory I heard.

**Comment by asubiotto** [comment]
I checked the memory accounting path and the initialization and tracking seem fine. Other queries hit expected budget limits in the `buildPhase` of the `hashJoiner` which confirmed this.

I've been looking at the specific reason for the OOM (namely a lot of `DDecimal` allocations). The latest hypothesis I had was that `DDecimal.Size` was not returning the correct size for a `DDecimal` and therefore tracking less memory than what was in use. The fact that we had `5.4GiB` used by this path, which is higher than the budget (`4.2GiB`) coupled with the fact that memory accounting seemed fine seemed to lend credibility to this. However, I got @mjibson to help me check this and it seems that `Size` is behaving correctly.

I'm kind of stumped on this and not sure how to proceed. I may be missing something obvious. One thing I am confused about is that `list DecodeTableValue` shows:

```
    5.29GB     5.29GB (flat, cum) 63.64% of Total
         .          .   1155:		// parser.DBoolTrue or parser.DBoolFalse.
         .          .   1156:		return parser.MakeDBool(parser.DBool(x)), b, err
         .          .   1157:	case parser.TypeInt:
         .          .   1158:		var i int64
         .          .   1159:		b, i, err = encoding.DecodeIntValue(b)
  359.54MB   359.54MB   1160:		return a.NewDInt(parser.DInt(i)), b, err
         .          .   1161:	case parser.TypeFloat:
         .          .   1162:		var f float64
         .          .   1163:		b, f, err = encoding.DecodeFloatValue(b)
         .          .   1164:		return a.NewDFloat(parser.DFloat(f)), b, err
         .          .   1165:	case parser.TypeDecimal:
         .          .   1166:		var d *apd.Decimal
         .        2MB   1167:		b, d, err = encoding.DecodeDecimalValue(b)
    2.72GB     2.72GB   1168:		dd := a.NewDDecimal(parser.DDecimal{})
    2.23GB     2.23GB   1169:		dd.Set(d)
         .          .   1170:		return dd, b, err
         .          .   1171:	case parser.TypeString:
         .          .   1172:		var data []byte
         .          .   1173:		b, data, err = encoding.DecodeBytesValue(b)
         .          .   1174:		return a.NewDString(parser.DString(data)), b, err
         .          .   1175:	case parser.TypeName:
         .          .   1176:		var data []byte
         .          .   1177:		b, data, err = encoding.DecodeBytesValue(b)
         .          .   1178:		return a.NewDName(parser.DString(data)), b, err
         .          .   1179:	case parser.TypeBytes:
         .          .   1180:		var data []byte
         .          .   1181:		b, data, err = encoding.DecodeBytesValue(b)
         .          .   1182:		return a.NewDBytes(parser.DBytes(data)), b, err
         .          .   1183:	case parser.TypeDate:
         .          .   1184:		var i int64
         .          .   1185:		b, i, err = encoding.DecodeIntValue(b)
         .          .   1186:		return a.NewDDate(parser.DDate(i)), b, err
         .          .   1187:	case parser.TypeTimestamp:
         .          .   1188:		var t time.Time
         .          .   1189:		b, t, err = encoding.DecodeTimeValue(b)
         .          .   1190:		return a.NewDTimestamp(parser.DTimestamp{Time: t}), b, err
         .          .   1191:	case parser.TypeTimestampTZ:
```

Where the memory is split between `NewDDecimal` and `Set` and I'm not sure why `Set` is an allocation since it simply sets values.

**Comment by RaduBerinde** [comment]
I think the lines may be offset by 1 and there is an allocation in `DecodeDecimalValue` and one in `NewDDecimal`. One possibility for the discrepancy is the use of `DatumAlloc` in `DecodeTableValue`. If say we filter out 4 in 5 values, and store only one, we still use all the memory for the slice. What's worse - the unused datums just sit in that slice so whatever they point to is kept live as well.

**Comment by RaduBerinde** [comment]
Ah, scratch that first part - `dd.Set()` allocates, it makes a copy of the value. I think we should just copy the `Decimal` instead of using `Set`. Also, `DecodeDecimalValue()` should return an `apd.Decimal` instead of allocating a `*apd.Decimal` @mjibson

**Comment by cuongdo** [comment]
This is a bad bug and is hampering further load testing. What's the latest here?

**Comment by asubiotto** [comment]
I stepped back from it for a while and haven't looked at it since. It might be good to get fresh eyes on this? I'm not sure when I'll be able to look at this again.

**Comment by RaduBerinde** [comment]
PR #15829 made some improvements to the way we allocate decimals, an updated profile might be more useful now.

**Comment by cuongdo** [comment]
@RaduBerinde there are a few ongoing efforts that help this, so can you ensure that this is fixed?

**Comment by asubiotto** [comment]
I have a hypothesis:
The `DDecimal`s are tracked by a `sqlbase.RowContainer` which keeps track of the memory used by these `Datums` correctly. However, there is some [preallocation](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/row_container.go#L200) of chunks done [when adding a row](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/row_container.go#L258) that doesn't track the [`Datum` references](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/row_container.go#L42). This memory use can grow quite large with the number of rows and is probably what caused an OOM.
The solution is not quite simple because it is not enough to track the preallocated `Datum` references. Under the hood, as soon as you [append to `chunks`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/row_container.go#L213) the actual memory taken up by `c.chunks` depends on how go decides to grow its arrays.
I will test this hypothesis and if proven true, the solution will be to account for the actual memory usage of `c.chunks` (which involves understanding how go does array expansion and mirroring that).

**Comment by RaduBerinde** [comment]
>  Under the hood, as soon as you append to chunks the actual memory taken up by c.chunks depends on how go decides to grow its arrays.

This is true, but note that `chunks` is just a slice of slices, and there is a slice for each chunk of rows. So the usage per-row is pretty small (a few bytes per row on average), even if we allow some headroom for array growth (in practice I think it's 1.25x for large arrays). There is no `Datum` storage unaccounted for, unless I'm missing something?

**Comment by RaduBerinde** [comment]
(Admittedly, I should have recorded this thought process in a comment).

**Comment by asubiotto** [comment]
You're right. It seems that `Datum` references are tracked even when the chunk is uninitialized so we do track the usage of `c.chunks` correctly (minus the headroom which is very small as you say). However the thing that is more important is that when growing an array, a copy is made, and that copy is not tracked.

Take for example the `5GB` that are in use in the memory profile in this issue and let's assume it's one of the first hashJoiners building the hash table with a 3-column row (`INT`, `DECIMAL`, `DECIMAL`). `rowSize` (used to determine the size of rows to report to the memory monitor) returns `184` bytes. This is around 29 million rows. `c.chunks` stores *references* to these datums. At 22 rows per chunk, we have `1,326,262` chunks in total (i.e. `len(c.chunks)`) and at a size of `1080` bytes per chunk, `c.chunks` takes up `1.4GB` in total.

Assume we are [appending to `c.chunks`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/row_container.go#L213) and trigger a growth of this slice. I checked and go arrays do grow by `1.25`. What happens here though is that a *new* array is created in another region of memory and the references to the `Datum`s are copied over but the old array is not gone until the next garbage collection run. Given that uninitialized chunks are `24` bytes, the memory usage is going to be:
```
numChunks = 1,326,262
memOriginal = numChunks * 1080 // Memory of original array
memNew = (numChunks * 1080) + ((numChunks * 0.25) * 24)) // Memory of new array
memOriginal + memNew == 2.8GB
```
That's `1.4GB` of unaccounted memory. This is what I *think* is going wrong but it's very possible I made a mistake.

**Comment by asubiotto** [comment]
I think I made a mistake. I missed an extra level of indirection. We don't copy over `1080` bytes per chunk since we don't copy all the `Datum` references, just the slice (so the pointer to the first `Datum`) in that slice which means ~`40MB` of unaccounted memory so I guess this doesn't really make sense.

**Comment by RaduBerinde** [comment]
Yes, that's what I meant by "slice of slices". Good point that growing by 1.25 means that we momentarily need 2.25x the memory for it. But it's still much smaller than the budgets we're talking about here.

**Comment by RaduBerinde** [comment]
I've been running this on scalefactor 10 on a 3 node cluster with `sql.defaults.distsql.tempstorage = false` and I saw no crashes. I did see occasional "budget exceeded" errors. Tried it with concurrency 3 and 6 for a few hours.

I'm closing this. Note that even if there is some accounting discrepancy, we are in much better shape with `tempstorage` enabled by default.
